source/h1_encoder.c

```diff
@@ -319,6 +344,309 @@ int aws_h1_encoder_start_message(
+                aws_last_error(),
+                aws_error_name(aws_last_error()));
+
+            return AWS_OP_ERR;
+        }
+
+        if ((amount_read > encoder->message->content_length) ||
+            (encoder->progress_bytes > encoder->message->content_length - amount_read)) {
+            ENCODER_LOGF(
+                ERROR, encoder, "Body stream has exceeded Content-Length: %" PRIu64, encoder->message->content_length);
+            return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+        }
+
+        encoder->progress_bytes += amount_read;
+
+        ENCODER_LOGF(TRACE, encoder, "Writing %zu body bytes to message", amount_read);
+
+        if (encoder->progress_bytes == encoder->message->content_length) {
+            ENCODER_LOG(TRACE, encoder, "Done sending body.");
+            encoder->progress_bytes = 0;
+            encoder->state++;
+            break;
+        }
+
+        /* Return if user failed to write anything. Maybe their data isn't ready yet. */
+        if (amount_read == 0) {
+            /* Ensure we're not at end-of-stream too early */
+            struct aws_stream_status status;
+            err = aws_input_stream_get_status(encoder->message->body, &status);
+            if (err) {
+                ENCODER_LOGF(
+                    TRACE,
+                    encoder,
+                    "Failed to query body stream status, error %d (%s)",
+                    aws_last_error(),
+                    aws_error_name(aws_last_error()));
+
+                return AWS_OP_ERR;
+            }
+            if (status.is_end_of_stream) {
+                ENCODER_LOGF(
+                    ERROR,
+                    encoder,
+                    "Reached end of body stream before Content-Length: %" PRIu64 " sent",
+                    encoder->message->content_length);
+                return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+            }
+
+            ENCODER_LOG(
+                TRACE,
+                encoder,
+                "No body data written, concluding this message. "
+                "Will try to write body data again in the next message.");
+            return AWS_OP_SUCCESS;
+        }
+    }
+    return AWS_OP_SUCCESS;
+}
+
+bool write_crlf(struct aws_byte_buf *dst) {
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    char crlf[CRLF_SIZE] = {'\r', '\n'};
+    struct aws_byte_cursor crlf_cursor = aws_byte_cursor_from_array(&crlf, AWS_ARRAY_SIZE(crlf));
+    return aws_byte_buf_write_from_whole_cursor(dst, crlf_cursor);
+}
+
+bool write_chunk_size(struct aws_byte_buf *dst, struct aws_input_stream *chunk) {
+    AWS_PRECONDITION(dst);
+    AWS_PRECONDITION(aws_byte_buf_is_valid(dst));
+    AWS_PRECONDITION(chunk);
+    AWS_PRECONDITION((dst->capacity - dst->len) >= MAX_ASCII_HEX_CHUNK_STR_SIZE);
+    int64_t chunk_size = 0;
+    if (AWS_OP_SUCCESS != aws_input_stream_get_length(chunk, &chunk_size)) {
```

Comment:
not all types of `aws_input_stream` support seeking, which the get_length() function relies on.
> a valid stream may return an error instead when there is not a good answer (socket stream, for example).

It would be more flexible to let the user pass the chunk size in through the options struct

Reply:
done. I added tests for incorrect stream length now that the user pass in an incorrect value.